crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,268 @@
+use wasm_bindgen::throw_str;
+
+pub mod callbacks {
+    use crate::{
+        blob::Blob,
+        file_reader::{FileReadError, ReadyState},
+    };
+    use gloo_events::EventListener;
+    use std::sync::{Arc, Mutex};
+    use wasm_bindgen::{throw_str, JsCast, UnwrapThrowExt};
+
+    /// A guard object that aborts the file read when dropped (if the read isn't already
+    /// finished).
+    #[derive(Debug)]
+    pub struct FileReader {
+        reader: web_sys::FileReader,
+        load_listener: EventListener,
+        error_listener: EventListener,
+    }
+
+    impl std::ops::Drop for FileReader {
+        fn drop(&mut self) {
+            if !ReadyState::from(self.reader.ready_state()).is_done() {
+                self.reader.abort();
+            }
+        }
+    }
+
+    /// Read the contents of `blob` into a text string, and then run the callback.
+    ///
+    /// If the returned `FileReader` is dropped before the callback is invoked, the read will be
+    /// cancelled.
+    pub fn read_as_text<F>(blob: &Blob, callback: F) -> FileReader
+    where
+        F: FnOnce(Result<String, FileReadError>) + 'static,
+    {
+        // we need to be able to run the FnOnce, while proving to the compiler that it can only run
+        // once. The easiest way is to `take` it out of an Option. The `Arc` and `Mutex` are
+        // because we need shared ownership and mutability (for the `take`).
+        let load_callback: Arc<
+            Mutex<Option<Box<dyn FnOnce(Result<String, FileReadError>) + 'static>>>,
```

Comment:
Does this need `Box`, or can it get away with storing it directly, like this?

```rust
let load_callback: Rc<RefCell<Option<F>>> = Rc::new(RefCell::new(Some(callback)));
```

Reply:
You are correct. :)